R/human_infection.R

```diff
@@ -180,53 +184,136 @@ infection_outcome_process <- function(
         timestep,
         parameters$ud
       )
+      
+      clinical <- calculate_clinical_infections(
+        variables,
+        infected_humans,
+        parameters,
+        renderer,
+        timestep
+      )
+      
+      treated <- calculate_treated(
+        variables,
+        clinical,
+        parameters,
+        timestep,
+        renderer
+      )
+      
+      update_severe_disease(
+        timestep,
+        infected_humans,
+        variables,
+        parameters,
+        renderer
+      )
+      
+      to_D <- treated$not(FALSE)$and(clinical)
+      to_A <- infected_humans$and(clinical$not(FALSE))
+      to_U <- NULL
+      
     } else if (parameters$parasite == "vivax"){
+      
       boost_immunity(
         variables$iaa,
         infected_humans,
         variables$last_boosted_iaa,
         timestep,
         parameters$ua
       )
+      
+      ## Only S and U infections are considered in generating lm-det infections
+      lm_detectable <- calculate_lm_det_infections(
+        variables,
+        variables$state$get_index_of(c("S","U"))$and(infected_humans),
+        parameters,
+        renderer,
+        timestep
+      )
+      
+      # Lm-detectable level infected S and U, and all A infections may receive clinical infections
+      # There is a different calculation to generate clinical infections, based on current infection level
+      # LM infections must only pass through the clinical calculation, therefore all "A" infections are included
+      # "S" and "U" infections must pass through the lm-detectable calculation prior to and in addition to the clinical
+      # calculation. We therefore consider all "A" infections and only the "S" and "U" infections that are now lm-detectable.
+      clinical <- calculate_clinical_infections(
+        variables,
+        variables$state$get_index_of("A")$and(infected_humans)$or(lm_detectable),
+        parameters,
+        renderer,
+        timestep
+      )
+      
+      treated <- calculate_treated(
+        variables,
+        clinical,
+        parameters,
+        timestep,
+        renderer
+      )
+      
+      to_U <- infected_humans$and(lm_detectable$not(F))$and(variables$state$get_index_of(c("S")))
+      to_A <- lm_detectable$and(clinical$not(F))
+      to_D <- clinical$and(treated$not(F))
```

Comment:
I wonder if a bitset is getting erroneously overwritten somewhere here?

if I add in this check 
``` 
      ih_s <- infected_humans$size()
      to_U <- infected_humans$and(lm_detectable$not(F))$and(variables$state$get_index_of(c("S")))
      to_A <- lm_detectable$and(clinical$not(F))
      to_D <- clinical$and(treated$not(F))
      if(to_D$size() + to_A$size() + to_U$size() != ih_s){
        browser()
      }
```
It gets triggered when I run the model.
I presume that that sum of people going to these states should = the number infected?

Reply:
It was similar in the Pf part too.

Reply:
I don't think this is quite right. You're right in that infected_humans, lm_detectable and clinical all get replaced when they get called by the new subset, but this should be ok because they aren't called again following their respective assignments.

The original infected_humans bitset also includes the "to_T" assignments, which get assigned elsewhere (in the calculate_treated function). Is this the cause of the issue?